src/libgit2/sparse.c

```diff
@@ -13,26 +13,124 @@
 #include "index.h"
 #include "ignore.h"
 
-static bool sparse_lookup_in_rules(
+#define HAS_FLAG(entity, flag) (((entity->flags & flag) != 0))
+
+static bool pattern_is_cone(const git_attr_fnmatch *match) {
+	size_t i;
+
+	if (match->length == 0) {
+		return false;
+	}
+
+  if (match->length == 1 && match->pattern[0] == '*') {
+		// NOTE(christoph): "/*" and "!/*/" both parse to "*", either to:
+		//   positive, wildcard, non-directory, or
+		//   negative, wildcard, directory.
+		bool is_dir = HAS_FLAG(match, GIT_ATTR_FNMATCH_DIRECTORY);
+		bool is_negative =  HAS_FLAG(match, GIT_ATTR_FNMATCH_NEGATIVE);
+		return is_dir == is_negative;
+  }
+
+	if (!HAS_FLAG(match, GIT_ATTR_FNMATCH_DIRECTORY)) {
+    return false;
+  };
+
+	if (HAS_FLAG(match, GIT_ATTR_FNMATCH_HASWILD)) {
+		for (i = 0; i < match->length - 1; i++) {
+			if (match->pattern[i] == '*') {
+				// NOTE(christoph): The ony acceptable wildcard is at the end.
+				return false;
+			}
+		}
+	}
+
+	return true;
+}
+
+static bool is_top_level_file(git_attr_path *path) {
+	return !path->is_dir && (strchr(path->path, '/') == NULL);
+}
+
+static bool pattern_matches_path(git_attr_fnmatch *match, git_attr_path *path, size_t path_length) {
+	// This is the number of characters that must match exactly.
+	// Ex: "A/B"  -> 3,
+	//     "A/B/C/*" -> 5
+	size_t exact_match_length = match->length;
+
+	size_t i;
+
+	// If we have a pattern like "A/B/C/*", we have to match "A/B/C" exactly, then have a slash,
+	// then have at least one more slash (or be a directory).
+	bool expected_extra_nesting = false;
+
+	if (HAS_FLAG(match, GIT_ATTR_FNMATCH_HASWILD)) {
+		expected_extra_nesting = true;
+		exact_match_length = match->length - 2; // Cut off the trailing "/*"
+	}
+
+	if (path_length < exact_match_length) {
+		return false;
+	}
+
+	if (expected_extra_nesting) {
+		if (path_length < exact_match_length + 2) {
+			return false;
+		}
+
+		if (!path->is_dir) {
+			bool found_slash = false;
+			for (i = exact_match_length + 1; i < path_length; i++) {
+				if (path->path[i] == '/') {
+					found_slash = true;
+					break;
+				}
+			}
+
+			if (!found_slash) {
+				return false;
+			}
+		}
+	}
+
+	if (path_length > exact_match_length && path->path[exact_match_length] != '/') {
+		return false;
+	}
+
+	for (i = 0; i < exact_match_length; i++) {
+		if (path->path[i] != match->pattern[i]) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static int sparse_lookup_in_rules(
         int *checkout,
         git_attr_file *file,
         git_attr_path *path)
 {
 	size_t j;
 	git_attr_fnmatch *match;
 	
+  int path_length = strlen(path->path);
+	if (is_top_level_file(path) ) {
+*checkout  =GIT_SPARSE_CHECKOUT;
```

Comment:
Looks like there is some tabs vs. spaces thing going on here leading to non-consistent whitespace.

Reply:
Fixed